multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mmrcms/MMRCMS.kt

```diff
@@ -1,529 +1,487 @@
-                FilterList(
-                    getInitialFilterList() + UriSelectFilter(
-                        "Tag",
-                        "tag",
+                add(Filter.Separator())
+
+                if (categories.isNotEmpty()) {
+                    add(
+                        UriPartFilter(
+                            "Category",
+                            "cat",
+                            arrayOf(
+                                "Any" to "",
+                                *categories.toTypedArray(),
+                            ),
+                        ),
+                    )
+                }
+
+                add(
+                    UriPartFilter(
+                        "Title begins with",
+                        "alpha",
                         arrayOf(
-                            "" to "Any",
-                            *tagMappings.toTypedArray(),
+                            "Any" to "",
+                            *"#ABCDEFGHIJKLMNOPQRSTUVWXYZ".toCharArray().map {
+                                Pair(it.toString(), it.toString())
+                            }.toTypedArray(),
                         ),
                     ),
                 )
+
+                if (tags.isNotEmpty()) {
+                    add(
+                        UriPartFilter(
+                            "Tag",
+                            "tag",
+                            arrayOf(
+                                "Any" to "",
+                                *tags.toTypedArray(),
+                            ),
+                        ),
+                    )
+                }
+
+                add(SortFilter())
             }
-            else -> FilterList(getInitialFilterList())
         }
+
+        return FilterList(filters)
     }
 
-    /**
-     * Class that creates a select filter. Each entry in the dropdown has a name and a display name.
-     * If an entry is selected it is appended as a query parameter onto the end of the URI.
-     * If `firstIsUnspecified` is set to true, if the first entry is selected, nothing will be appended on the the URI.
-     */
-    // vals: <name, display>
-    open class UriSelectFilter(
-        displayName: String,
-        private val uriParam: String,
-        private val vals: Array<Pair<String, String>>,
-        private val firstIsUnspecified: Boolean = true,
-        defaultValue: Int = 0,
-    ) :
-        Filter.Select<String>(displayName, vals.map { it.second }.toTypedArray(), defaultValue), UriFilter {
-        override fun addToUri(uri: Uri.Builder) {
-            if (state != 0 || !firstIsUnspecified) {
-                uri.appendQueryParameter(uriParam, vals[state].first)
+    private var categories = emptyList<Pair<String, String>>()
+
+    private var statuses = emptyList<Pair<String, String>>()
+
+    private var types = emptyList<Pair<String, String>>()
+
+    private var tags = emptyList<Pair<String, String>>()
```

Comment:
Could you combine these into a single `HashMap`? Plus, is it possible to infer the filters from the page instead of hardcoding them?

Reply:
Edit: A regular map for insert order, sorry.

Reply:
Never mind, I actually mean if the filters have the same HTML structure, but it looks like your logic is more complicated.